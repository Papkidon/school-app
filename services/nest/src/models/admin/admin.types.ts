import { z } from 'zod';

export const adminCreateSchema = z.object({
  data: z.object({
    username: z.string().min(2).max(20),
    password: z.string().min(8).max(20),
  }),
});

export const adminUpdateSchema = adminCreateSchema.deepPartial();

export type adminType = {
  data: {
    username: string;
    password: string;
  };
};
