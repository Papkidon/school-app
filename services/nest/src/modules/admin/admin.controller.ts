import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import {
  adminCreateSchema,
  adminUpdateSchema,
} from '../../models/admin/admin.types';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { Admin } from '@prisma/client';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { AdminService } from './admin.service';

@Controller('api/v1/admin')
export class AdminController {
  constructor(private readonly adminService: AdminService) {}

  @Get('')
  async findAll(): Promise<Admin[]> {
    try {
      const result = await this.adminService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No admin data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(adminCreateSchema)) createAdminDto: adminType,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findByUsername(
        createAdminDto.data.username,
      );

      if (userExists) {
        throw new HttpException(
          'User with given username already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.adminService.create(createAdminDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Admin> {
    try {
      const result = await this.adminService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(adminUpdateSchema))
    body: DeepPartial<adminType>,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'Admin with given id does not exist',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.adminService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'Admin with given id does not exist',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.adminService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
