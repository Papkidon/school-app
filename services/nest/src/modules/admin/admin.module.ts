import { Module } from '@nestjs/common';
import { PrismaService } from 'nestjs-prisma';
import { AdminController } from './admin.controller';
import { AdminRepository } from './admin.repository';
import { AdminService } from './admin.service';

@Module({
  controllers: [AdminController],
  providers: [AdminService, AdminRepository, PrismaService],
})
export class AdminModule {}
