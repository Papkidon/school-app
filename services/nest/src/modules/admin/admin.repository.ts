import { Injectable } from '@nestjs/common';
import { PrismaService as PrismaService } from 'nestjs-prisma';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial } from '../../models/utility.types';
import { Admin } from '@prisma/client';

@Injectable()
export class AdminRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Admin[]> {
    return await this.prisma.admin.findMany();
  }

  async findById(id: string): Promise<Admin> {
    return await this.prisma.admin.findUnique({ where: { id: id } });
  }

  async create(body: any): Promise<Admin> {
    return await this.prisma.admin.create(body);
  }

  async delete(id: string): Promise<Admin> {
    return await this.prisma.admin.delete({ where: { id: id } });
  }

  async update(id: string, body: DeepPartial<adminType>): Promise<Admin> {
    return await this.prisma.admin.update({
      where: { id: id },
      data: { ...body.data },
    });
  }
}
