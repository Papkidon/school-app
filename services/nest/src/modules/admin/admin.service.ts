import { Injectable } from '@nestjs/common';
import { AdminRepository } from './admin.repository';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial } from '../../models/utility.types';
import { Admin } from '@prisma/client';

@Injectable()
export class AdminService {
  constructor(private adminRepository: AdminRepository) {}
  async findAll(): Promise<Admin[]> {
    return await this.adminRepository.findAll();
  }

  async findById(id: string): Promise<Admin> {
    return await this.adminRepository.findById(id);
  }

  async create(body: any): Promise<Admin> {
    return await this.adminRepository.create(body);
  }

  async delete(id: string): Promise<Admin> {
    return await this.adminRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<adminType>): Promise<Admin> {
    return await this.adminRepository.update(id, body);
  }
}
