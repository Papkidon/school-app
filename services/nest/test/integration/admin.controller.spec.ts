import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { AdminModule } from '../../src/modules/admin/admin.module';
import { AdminRepository } from '../../src/modules/admin/admin.repository';
import { dummyAdminCreate } from './dummies/dummy.data';

describe('[AdminController] Integration tests', () => {
  let app: INestApplication;
  const adminRepository = {
    findAll: () => ['dummyFoundAdmin'],
    findById: () => 'dummyFoundAdminById',
    findByUsername: () => 'dummyFoundAdminByUsername',
    create: () => 'dummyCreatedAdmin',
    update: () => 'dummyUpdatedAdmin',
    delete: () => 'dummyDeletedAdmin',
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [AdminModule],
    })
      .overrideProvider(AdminRepository)
      .useValue(adminRepository)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it(`Get all admins - should pass when at least one admin exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get('/api/v1/admin/');

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get all admins - should fail when no admin data exists in the db`, async () => {
    jest.spyOn(adminRepository, 'findAll').mockImplementationOnce(() => []);

    const response = await request(app.getHttpServer()).get('/api/v1/admin/');

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get admin by id - should pass when admin with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get admin by id - should fail when admin with given id does not exists in the db`, async () => {
    jest.spyOn(adminRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).get(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it('Get admin by id - should fail when id parameter is not UUID', async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/admin/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create admin - should pass when request body is correct`, async () => {
    jest
      .spyOn(adminRepository, 'findByUsername')
      .mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer())
      .post('/api/v1/admin/')
      .send(dummyAdminCreate);

    expect(response.status).toEqual(201);
    expect(response.body).toMatchSnapshot();
  });

  it('Create admin - should fail when request body is incorrect', async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/admin/')
      .send({ wrong: 'body' });

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update admin - should pass when admin with given id exists in the db`, async () => {
    jest.spyOn(adminRepository, 'findById').mockImplementationOnce(() => 'xD');
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update admin - should fail when admin with given id does not exists in the db`, async () => {
    jest.spyOn(adminRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update admin - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/admin/123',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete admin - should pass when admin with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete admin - should fail when admin with given id does not exist in the db`, async () => {
    jest.spyOn(adminRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).delete(
      '/api/v1/admin/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete admin - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/admin/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  afterAll(async () => {
    await app.close();
  });
});
