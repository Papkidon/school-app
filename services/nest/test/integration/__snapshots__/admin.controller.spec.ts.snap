// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AdminController] Integration tests Create admin - should fail when request body is incorrect 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Required",
    "path": [
      "data",
    ],
    "received": "undefined",
  },
  {
    "code": "unrecognized_keys",
    "keys": [
      "wrong",
    ],
    "message": "Unrecognized key(s) in object: 'wrong'",
    "path": [],
  },
]
`;

exports[`[AdminController] Integration tests Create admin - should pass when request body is correct 1`] = `{}`;

exports[`[AdminController] Integration tests Delete admin - should fail when admin with given id does not exist in the db 1`] = `
{
  "message": "Admin with given id does not exist",
  "statusCode": 404,
}
`;

exports[`[AdminController] Integration tests Delete admin - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[AdminController] Integration tests Delete admin - should pass when admin with given id exists in the db 1`] = `{}`;

exports[`[AdminController] Integration tests Get admin by id - should fail when admin with given id does not exists in the db 1`] = `
{
  "message": "User with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[AdminController] Integration tests Get admin by id - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[AdminController] Integration tests Get admin by id - should pass when admin with given id exists in the db 1`] = `{}`;

exports[`[AdminController] Integration tests Get all admins - should fail when no admin data exists in the db 1`] = `
{
  "message": "No admin data exists in the database",
  "statusCode": 404,
}
`;

exports[`[AdminController] Integration tests Get all admins - should pass when at least one admin exists in the db 1`] = `
[
  "dummyFoundAdmin",
]
`;

exports[`[AdminController] Integration tests Update admin - should fail when admin with given id does not exists in the db 1`] = `
{
  "message": "Admin with given id does not exist",
  "statusCode": 404,
}
`;

exports[`[AdminController] Integration tests Update admin - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[AdminController] Integration tests Update admin - should pass when admin with given id exists in the db 1`] = `{}`;
